File: jlox-in-cpp/Scanner.cpp
#include "Scanner.h"

#include <charconv>
#include <cstdlib>
#include <unordered_map>

#include "Error.h"

const std::vector<Token> &Scanner::scanTokens() {
  while (!isAtEnd()) {
    // We are at the beginning of the next lexeme.
    start = current;
    scanToken();
  }

  tokens.emplace_back(TokenType::TOKEN_EOF, "", line);
  return tokens;
}

void Scanner::scanToken() {
  char c = advance();
  switch (c) {
  case '(':
    addToken(TokenType::LEFT_PAREN);
    break;
  case ')':
    addToken(TokenType::RIGHT_PAREN);
    break;
  case '{':
    addToken(TokenType::LEFT_BRACE);
    break;
  case '}':
    addToken(TokenType::RIGHT_BRACE);
    break;
  case ',':
    addToken(TokenType::COMMA);
    break;
  case '.':
    addToken(TokenType::DOT);
    break;
  case '-':
    addToken(TokenType::MINUS);
    break;
  case '+':
    addToken(TokenType::PLUS);
    break;
  case ';':
    addToken(TokenType::SEMICOLON);
    break;
  case '*':
    addToken(TokenType::STAR);
    break;
  case '!':
    addToken(match('=') ? TokenType::BANG_EQUAL : TokenType::BANG);
    break;
  case '=':
    addToken(match('=') ? TokenType::EQUAL_EQUAL : TokenType::EQUAL);
    break;
  case '<':
    addToken(match('=') ? TokenType::LESS_EQUAL : TokenType::LESS);
    break;
  case '>':
    addToken(match('=') ? TokenType::GREATER_EQUAL : TokenType::GREATER);
    break;

  case '/':
    if (match('/')) {
      // A comment goes until the end of the line.
      while (peek() != '\n' && !isAtEnd())
        advance();
    } else {
      addToken(TokenType::SLASH);
    }
    break;

  case ' ':
  case '\r':
  case '\t':
    // Ignore whitespace.
    break;

  case '\n':
    ++line;
    break;

  case '"':
    string();
    break;

  default:
    if (std::isdigit(c))
      number();
    else if (std::isalpha(c) || c == '_')
      identifier();
    else
      error(line, "Unexpected character.");
    break;
  }
}

void Scanner::identifier() {
  static const std::unordered_map<std::string_view, TokenType> keywords = {
      {"and", TokenType::AND},       {"class", TokenType::CLASS},
      {"else", TokenType::ELSE},     {"false", TokenType::FALSE},
      {"for", TokenType::FOR},       {"fun", TokenType::FUN},
      {"if", TokenType::IF},         {"nil", TokenType::NIL},
      {"or", TokenType::OR},         {"print", TokenType::PRINT},
      {"return", TokenType::RETURN}, {"super", TokenType::SUPER},
      {"this", TokenType::THIS},     {"true", TokenType::TRUE},
      {"var", TokenType::VAR},       {"while", TokenType::WHILE},
  };

  auto isAlphaNumeric = [](char c) {
    return std::isalpha(c) || std::isdigit(c) || c == '_';
  };

  while (isAlphaNumeric(peek()))
    advance();

	auto it = keywords.find(currentLexeme());
  if (it == keywords.end())
    addToken(TokenType::IDENTIFIER);
  else
    addToken(it->second);
}

void Scanner::number() {
  while (std::isdigit(peek()))
    advance();

  // Look for a fractional part.
  if (peek() == '.' && std::isdigit(peekNext())) {
    // Consume the "."
    advance();

    while (std::isdigit(peek()))
      advance();
  }

  // libc++ doesn't implement decimal from_chars yet, and strtod doesn't accept
  // an end point, so we have to manually copy it out.
  std::string numberString(&source.at(start), current - start);
  addToken(std::strtod(numberString.data(), nullptr));
}

void Scanner::string() {
  while (peek() != '"' && !isAtEnd()) {
    if (peek() == '\n')
      ++line;
    advance();
  }

  if (isAtEnd()) {
    error(line, "Unterminated string.");
    return;
  }

  // The closing ".
  advance();

  // Trim the surrounding quotes.
  addToken(source.substr(start + 1, current - start - 2));
}

bool Scanner::match(char expected) {
  if (isAtEnd())
    return false;
  if (peek() != expected)
    return false;

  ++current;
  return true;
}

char Scanner::peek() {
  if (isAtEnd())
    return '\0';
  return source.at(current);
}

char Scanner::peekNext() {
  if (current + 1 >= source.length())
    return '\0';
  return source.at(current + 1);
}

bool Scanner::isAtEnd() { return current >= source.length(); }

char Scanner::advance() { return source.at(current++); }

void Scanner::addToken(TokenType type) {
  tokens.emplace_back(type, currentLexeme(), line);
}

void Scanner::addToken(std::string_view literal) {
  tokens.emplace_back(TokenType::STRING, currentLexeme(), literal, line);
}

void Scanner::addToken(double literal) {
  tokens.emplace_back(TokenType::NUMBER, currentLexeme(), literal, line);
}

std::string_view Scanner::currentLexeme() {
  return source.substr(start, current - start);
}

File: jlox-in-rust/src/scanner.rs
use crate::error;
use crate::token::Token;
use crate::token_type::TokenType;

pub struct Scanner<'a> {
    // All the tokens we're scanning for are ASCII, so we can iterate much more simply and
    // efficiently over bytes. UTF-8 means that we'll still handle non-ASCII characters correctly.
    source: &'a [u8],
    source_str: &'a str,
    start: usize,
    current: usize,
    line: u32,
}

impl<'a> Scanner<'a> {
    pub fn new(source: &'a str) -> Self {
        Scanner {
            source: source.as_bytes(),
            source_str: source,
            start: 0,
            current: 0,
            line: 0,
        }
    }

    pub fn scan_tokens(&mut self) -> Vec<Token<'a>> {
        let mut tokens = Vec::new();
        while !self.is_at_end() {
            // We are at the beginning of the next lexeme.
            self.start = self.current;
            if let Some(token) = self.scan_token() {
                tokens.push(token);
            }
        }

        tokens.push(Token {
            r#type: TokenType::Eof,
            lexeme: "",
            line: 0,
        });
        tokens
    }

    fn scan_token(&mut self) -> Option<Token<'a>> {
        let c = self.advance();
        match c {
            b'(' => Some(self.add_token(TokenType::LeftParen)),
            b')' => Some(self.add_token(TokenType::RightParen)),
            b'{' => Some(self.add_token(TokenType::LeftBrace)),
            b'}' => Some(self.add_token(TokenType::RightBrace)),
            b',' => Some(self.add_token(TokenType::Comma)),
            b'.' => Some(self.add_token(TokenType::Dot)),
            b'-' => Some(self.add_token(TokenType::Minus)),
            b'+' => Some(self.add_token(TokenType::Plus)),
            b';' => Some(self.add_token(TokenType::Semicolon)),
            b'*' => Some(self.add_token(TokenType::Star)),

            b'!' => {
                let token_type = if self.r#match(b'=') {
                    TokenType::BangEqual
                } else {
                    TokenType::Bang
                };
                Some(self.add_token(token_type))
            }

            b'=' => {
                let token_type = if self.r#match(b'=') {
                    TokenType::EqualEqual
                } else {
                    TokenType::Equal
                };
                Some(self.add_token(token_type))
            }

            b'<' => {
                let token_type = if self.r#match(b'=') {
                    TokenType::LessEqual
                } else {
                    TokenType::Less
                };
                Some(self.add_token(token_type))
            }

            b'>' => {
                let token_type = if self.r#match(b'=') {
                    TokenType::GreaterEqual
                } else {
                    TokenType::Greater
                };
                Some(self.add_token(token_type))
            }

            b'/' => {
                if self.r#match(b'/') {
                    // A comment goes until the end of the line.
                    while self.peek() != b'\n' && !self.is_at_end() {
                        self.advance();
                    }
                    None
                } else {
                    Some(self.add_token(TokenType::Slash))
                }
            }

            b' ' | b'\r' | b'\t' => None,

            b'\n' => {
                self.line += 1;
                None
            }

            b'"' => self.string(),

            b'0'..=b'9' => self.number(),

            b'A'..=b'Z' | b'a'..=b'z' | b'_' => self.identifier(),

            _ => {
                error::error(self.line, "Unexpected character .");
                None
            }
        }
    }

    fn identifier(&mut self) -> Option<Token<'a>> {
        while let b'0'..=b'9' | b'A'..=b'Z' | b'a'..=b'z' | b'_' = self.peek() {
            self.advance();
        }

        let text = &self.source_str[self.start..self.current];
        let token_type = match text {
            "and" => TokenType::And,
            "class" => TokenType::Class,
            "else" => TokenType::Else,
            "false" => TokenType::False,
            "for" => TokenType::For,
            "fun" => TokenType::Fun,
            "if" => TokenType::If,
            "nil" => TokenType::Nil,
            "or" => TokenType::Or,
            "print" => TokenType::Print,
            "return" => TokenType::Return,
            "super" => TokenType::Super,
            "this" => TokenType::This,
            "true" => TokenType::True,
            "var" => TokenType::Var,
            "while" => TokenType::While,
            _ => TokenType::Identifier,
        };
        Some(self.add_token(token_type))
    }

    fn number(&mut self) -> Option<Token<'a>> {
        while let b'0'..=b'9' = self.peek() {
            self.advance();
        }

        // Look for a fractional part.
        if self.peek() == b'.' && self.peek_next().is_ascii_digit() {
            // Consume the "."
            self.advance();

            while let b'0'..=b'9' = self.peek() {
                self.advance();
            }
        }

        let number: f64 = self.source_str[self.start..self.current].parse().unwrap();
        Some(self.add_token(TokenType::Number(number)))
    }

    fn string(&mut self) -> Option<Token<'a>> {
        while self.peek() != b'"' && !self.is_at_end() {
            if self.peek() == b'\n' {
                self.line += 1;
            }
            self.advance();
        }

        if self.is_at_end() {
            error::error(self.line, "Unterminated string.");
            return None;
        }

        // The closing ".
        self.advance();

        // Trim the surrounding quotes.
        let value = &self.source_str[self.start + 1..self.current - 1];
        Some(self.add_token(TokenType::String(value)))
    }

    fn r#match(&mut self, expected: u8) -> bool {
        if self.is_at_end() {
            return false;
        }

        if self.source[self.current] != expected {
            return false;
        }

        self.current += 1;
        true
    }

    fn peek(&self) -> u8 {
        self.source.get(self.current).copied().unwrap_or(0)
    }

    fn peek_next(&self) -> u8 {
        self.source.get(self.current + 1).copied().unwrap_or(0)
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }

    fn advance(&mut self) -> u8 {
        let c = self.source[self.current];
        self.current += 1;
        c
    }

    fn add_token(&self, token_type: TokenType<'a>) -> Token<'a> {
        Token {
            r#type: token_type,
            lexeme: &self.source_str[self.start..self.current],
            line: self.line,
        }
    }
}
